### PR TITLE
✨ Add the ability to specify a drain timeout for machines

### LINKED ISSUE #2331: Allow a user specifiable node draining timeout
As a user/operator, I would like the option to specify a timeout for how long the machine controller will wait for a node to drain before forcibly removing the machine.

Today, if a user has specified a pod disruption budget that cannot be fulfilled, CAPI could wait forever on a machine/node that is running a pod whose budget could not be met and will never successfully drain. I would like the option to say ...

1. Wait X minutes for a node to drain.
2. If the node does not drain within that amount of time, continue on with the deletion anyways.

This would prevent CAPI from getting suck in an infinite reconcile loop during certain cluster conditions.

/kind feature

### PR DESCRIPTION
**Pull request by namnx228**
Add an option `nodeDrainTimeout` to KCP and machiceSpec of machinedeployment. 
`nodeDrainTimeout` defines the amount of time we want a node to be drained. The node is forcefully removed if the time is over.
Note: Unset this option means there is no time limit.

How does this PR do?
- Add an option `nodeDrainTimeout` to KCP and to MachineSpec field of Machideployment.
- KCP or Machinedeployment create machines which have this `nodeDrainTimeout` option.
- For each machine, In the first time the machine is drained, the conditions `DrainingSucceededCondition` is set to "False", and its last trasition time is the first time the machine is drained.
- The machine ignores the draining phase if the timeout is over.

Fixes #2331

### CODE REVIEW

**Comment by namnx228** [comment]
Signed CLA

**Comment by namnx228** [comment]
signed

**Comment by namnx228** [comment]
signed CLA

**Comment by namnx228** [comment]
signed CLA

File: api/v1alpha3/machine_types.go
Comment (vincepri) [inline api/v1alpha3/machine_types.go:93]:
```suggestion
	// NodeDrainTimeout is the total amount of time that the controller will spend on draining a node.
```

File: api/v1alpha3/machine_types.go
Comment (vincepri) [inline api/v1alpha3/machine_types.go:94]:
```suggestion
	//
	// NOTE: NodeDrainTimeout is different from `kubectl drain --timeout`
```

Let's also add some color on how it's different?

File: controllers/machine_controller.go
Comment (vincepri) [inline controllers/machine_controller.go:302]: Instead of chaining if conditions here, let's add a new if condition after entering this block that checks the timeout

File: api/v1alpha3/machine_types.go
Comment (vincepri) [reply api/v1alpha3/machine_types.go:94]: Should we have a default here?

File: controllers/machine_controller.go
Comment (vincepri) [inline controllers/machine_controller.go:312]: Can we add a comment above here that explains why we don't want to call `MarkFalse` and we should check if the condition exists?

I assumed during review this was to avoid the change to Condition's TransitionTime, it'd be good to make it clear

File: controllers/machine_controller.go
Comment (vincepri) [inline controllers/machine_controller.go:383]: Make this a method? Seems everything else is on the MachineReconcile object

File: controllers/machine_controller.go
Comment (vincepri) [inline controllers/machine_controller.go:370]:
```suggestion
func isNodeDrainAllowed(machine *clusterv1.Machine) bool {
```

To make it a little bit more generic, in case in the future we might want to check more than just a timeout. We can also add the check for the annotation `if _, exists := m.ObjectMeta.Annotations[clusterv1.ExcludeNodeDrainingAnnotation]`

**Review by vincepri — COMMENTED** [review]
Overall this looks good, thanks for working on it!

We'll need to add more tests and integration tests in the Machine controller, especially around the logic that checks if the node drain is still allowed or not

**Comment by namnx228** [comment]
Thanks for your review, @vincepri. I will push a fix that responds to your comments soon.
I have two questions: 
> 
> We'll need to add more tests and integration tests in the Machine controller, especially around the logic that checks if the node drain is still allowed or not
- Do we need the test to be included in the current PR, or we can have a separated PR for the test?
- It seems like the two failing CI tests are related to the API change. Should I ignore them?

**Comment by vincepri** [comment]
> Do we need the test to be included in the current PR, or we can have a separated PR for the test?

Same PR, usually changes should be tested before merging them, it has happened in the past that we've merged code without tests and forgot to follow-up, since then we've decided that PR should come with tests + docs whenever possible

> It seems like the two failing CI tests are related to the API change. Should I ignore them?

For now yes, if they keep failing after a number of retries we might want to look into it a bit more. We've merged a bunch of fixes for the tests, can you try rebasing?

File: api/v1alpha3/machine_types.go
Comment (CecileRobertMichon) [inline api/v1alpha3/machine_types.go:97]: I wonder if it would make more sense to add this to `MachineDeploymentStrategy` instead (and then to KCP separately) https://github.com/CecileRobertMichon/cluster-api/blob/2b71eebdec7e361e9baeb8f1454f8a6e0268c169/api/v1alpha3/machinedeployment_types.go#L101

What's the reasoning for setting this value for each machine individually?

Comment (namnx228) [reply api/v1alpha3/machine_types.go:97]: Hi @CecileRobertMichon, thanks for your review.
We actually don't set this value for each machine individually, but setting in the `machineSpec` of  `machinedeployment`. 
For example: 

```
apiVersion: cluster.x-k8s.io/v1alpha3
kind: MachineDeployment
metadata:
...
spec:
...
  template:
    metadata:
      labels:
        cluster.x-k8s.io/cluster-name: my-cluster
        nodepool: nodepool-0
    spec:
      clusterName: my-cluster
      version: v1.15.3
      bootstrap:
        ...
      infrastructureRef:
        ...
      nodeDrainTimeout: 180 

```

**Comment by namnx228** [comment]
The PR is failing to some API tests. The problem is the conversion from v1alpha3 to v1alpha2. My plan to solve this problem is to ignore the new field `NodeDrainTimeout` when conversing to v1alpha2, and we suppose that there is no option to set the node draining timeout in v1alpha2. How do you think about this approach @vincepri @CecileRobertMichon? 

Beside, do we have any documentations that helps to write integration tests? I know that the integration test directory is `cluster-api/test/e2e`, but a little more instructions is appreciated.

File: api/v1alpha2/conversion.go
Comment (vincepri) [inline api/v1alpha2/conversion.go:147]: This doesn't look right, we'd lose the information if we zero out the Timeout value during a conversion. Instead, we should restore the old value in the `restoreMachineSpec` function

File: api/v1alpha3/machine_types.go
Comment (CecileRobertMichon) [reply api/v1alpha3/machine_types.go:97]: Right, but it's still part of the template instead of being part of the MachineDeployment spec directly

Comment (ncdc) [reply api/v1alpha3/machine_types.go:97]: Draining is a per-node (machine) operation, though. Also look at the Pod spec - it has e.g. `terminationGracePeriodSeconds`, which is a similar type of thing to nodeDrainTimeout.

File: api/v1alpha2/conversion.go
Comment (namnx228) [reply api/v1alpha2/conversion.go]: Fixed! Thanks.

**Comment by namnx228** [comment]
> Overall this looks good, thanks for working on it!
> 
> We'll need to add more tests and integration tests in the Machine controller, especially around the logic that checks if the node drain is still allowed or not

@vincepri  I have added the unit tests in the Machine controller to test whether a node is allowed to drain or not. 
To test the integration between components, it is necessary to have workloads in the workload cluster that blocks the pod eviction process (using `podDisruptionBudget` to prevent some pods from being evicted). This prevention makes the node to be unable to be drained in time, resulting in a forceful deletion. In this case, the integration test cannot make it because it doesn't deploy a real workload cluster.  For that reason, e2e tests may be more suitable in this case. What do you think about this?

**Comment by vincepri** [comment]
> For that reason, e2e tests may be more suitable in this case. What do you think about this?

That sounds good to me, we have different e2e tests for this, we can add a new test case for it. We can follow-up in a different PR, I'll redirect you to @fabriziopandini and @sedefsavas to understand how to wire up a new test

**Comment by vincepri** [comment]
/retitle ✨ Add the ability to specify a drain timeout for machines

File: controllers/machine_controller.go
Comment (vincepri) [inline controllers/machine_controller.go:385]:
```suggestion
func (r *MachineReconciler) isNodeDrainTimeoutOver(machine *clusterv1.Machine) bool {
```

Comment (namnx228) [reply controllers/machine_controller.go]: Fixed! Thanks.

**Comment by namnx228** [comment]
>  We can follow-up in a different PR
Great. At the moment, can this PR get lgtm label?
> I'll redirect you to @fabriziopandini and @sedefsavas to understand how to wire up a new e2e test case

Nice I'd ask them when I need help. Thanks for letting me know

**Comment by namnx228** [comment]
> @namnx228: The following test **failed**, say `/retest` to rerun all failed tests:
> Test name 	Commit 	Details 	Rerun command
> pull-cluster-api-verify-external-links 	[6239da5](https://github.com/kubernetes-sigs/cluster-api/commit/6239da57bb887c20f9db220b53c657df20d66bfb) 	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/3662/pull-cluster-api-verify-external-links/1309032648298991616) 	`/test pull-cluster-api-verify-external-links`
> 
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api&pr=3662). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Anamnx228). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api/issues?q=is:issue+is:open) when you hit one in your PR.

Is it an issue in CI that the `K8s-CI-robot` keeps showing the failed test result of an old commit?

**Comment by vincepri** [comment]
@namnx228 You can ignore the `pull-cluster-api-verify-external-links`, it was moved to a periodic and it shouldn't actually run anymore

**Review by vincepri — APPROVED** [review]
/approve
/assign @CecileRobertMichon @ncdc

File: api/v1alpha3/machine_types.go
Comment (ncdc) [reply api/v1alpha3/machine_types.go:97]: @CecileRobertMichon ok to mark resolved?

Comment (ncdc) [inline api/v1alpha3/machine_types.go:97]: Shouldn't this be a metav1.Duration?

Comment (CecileRobertMichon) [reply api/v1alpha3/machine_types.go:97]: yes, marking as resolved 👍

Comment (namnx228) [reply api/v1alpha3/machine_types.go:97]: Yes, it can also be metav1.Duration. 
Do we have any significant advantage if now we change it from `int64` to `metav1.Duration` since it can lead to quite a lot of change to this PR?

File: controllers/machine_controller_test.go
Comment (ncdc) [inline controllers/machine_controller_test.go:810]: Do you need this?

File: api/v1alpha3/machine_types.go
Comment (ncdc) [reply api/v1alpha3/machine_types.go:97]: We typically use Duration for ... durations. The significant advantage is getting this right early on instead of having to make an API change later. @vincepri @CecileRobertMichon @detiber WDYT?

File: controllers/machine_controller.go
Comment (vincepri) [reply controllers/machine_controller.go:385]: SGTM, was struggling to find a better name the other day

File: api/v1alpha3/machine_types.go
Comment (vincepri) [reply api/v1alpha3/machine_types.go:97]: Yes please, that's a good point. We should try to be consistent with the rest of the codebase

Comment (namnx228) [reply api/v1alpha3/machine_types.go]: I see. Fixed!

File: controllers/machine_controller.go
Comment (namnx228) [reply controllers/machine_controller.go]: Fixed!

File: controllers/machine_controller.go
Comment (namnx228) [reply controllers/machine_controller.go]: Fixed! Thanks

File: controllers/machine_controller.go
Comment (namnx228) [reply controllers/machine_controller.go]: Changed! Thanks for the suggestion

File: controllers/machine_controller.go
Comment (namnx228) [reply controllers/machine_controller.go]: Fixed! Thanks

File: controllers/machine_controller_test.go
Comment (namnx228) [reply controllers/machine_controller_test.go]: Yes, we don't need it. Deleted. Thanks

**Comment by namnx228** [comment]
@ncdc all the required changes have been made. Please take a look again. Thanks!

**Comment by ncdc** [comment]
/lgtm

@namnx228 thanks for your patience w/the reviews, and thanks for doing this!

**Comment by namnx228** [comment]
@ncdc @vincepri Thank you very much for your review and suggestions. I will open a follow-up PR to provide e2e tests for this PR soon.

**Pull request merged**
